src/main/scala/zio/s3/S3.scala

```diff
@@ -0,0 +1,395 @@
+        .map(S3BucketListing(_))
+
+    override def getObject(bucketName: String, key: String): ZStreamChunk[Any, S3Exception, Byte] =
+      ZStreamChunk(
+        ZStream
+          .fromEffect(
+            execute(
+              _.getObject[StreamResponse](
+                GetObjectRequest.builder().bucket(bucketName).key(key).build(),
+                StreamAsyncResponseTransformer(new CompletableFuture[StreamResponse]())
+              )
+            )
+          )
+          .flatMap(identity)
+          .mapError(S3.S3ExceptionLike)
+      )
+
+    override def listObjects(bucketName: String, prefix: String, maxKeys: Int): ZIO[Any, S3Exception, S3ObjectListing] =
+      execute(
+        _.listObjectsV2(ListObjectsV2Request.builder().maxKeys(maxKeys).bucket(bucketName).prefix(prefix).build())
+      ).map(S3ObjectListing(_))
+
+    override def getNextObjects(listing: S3ObjectListing): ZIO[Any, S3Exception, S3ObjectListing] =
+      listing.nextContinuationToken
+        .fold[ZIO[Any, S3Exception, S3ObjectListing]](
+          ZIO.succeed(listing.copy(nextContinuationToken = None, objectSummaries = Nil))
+        ) { token =>
+          execute(
+            _.listObjectsV2(ListObjectsV2Request.builder().bucket(listing.bucketName).continuationToken(token).build())
+          ).map(S3ObjectListing(_))
+        }
+
+    override def putObject[R1](
+      bucketName: String,
+      key: String,
+      contentLength: Long,
+      contentType: String,
+      content: ZStreamChunk[R1, Throwable, Byte]
+    ): ZIO[Any with R1, S3Exception, Unit] =
+      content.chunks
+        .map(c => ByteBuffer.wrap(c.toArray))
+        .toPublisher
+        .flatMap(publisher =>
+          execute(
+            _.putObject(
+              PutObjectRequest
+                .builder()
+                .bucket(bucketName)
+                .contentLength(contentLength)
+                .contentType(contentType)
+                .key(key)
+                .build(),
+              AsyncRequestBody.fromPublisher(publisher)
+            )
+          )
+        )
+        .unit
+
+    // only for file which are bigger than 5 Mb
+    def multipartUpload[R1 <: Any](n: Int)(
+      bucketName: String,
+      key: String,
+      contentType: String,
+      content: ZStream[R1, Throwable, Byte]
+    ): ZIO[R1, S3Exception, Unit] =
+      for {
+        uploadId <- execute(
+                     _.createMultipartUpload(
+                       CreateMultipartUploadRequest
+                         .builder()
+                         .bucket(bucketName)
+                         .key(key)
+                         .contentType(contentType)
+                         .build()
+                     )
+                   ).map(_.uploadId())
+
+        parts <- content
+                //part size limit is 5Mb, required by amazon api
+                  .chunkN(5 * 1024 * 1024)
+                  .chunks
+                  .zipWithIndex
+                  .mapMPar(n) {
```

Comment:
TODO: Talk to Itamar: 

```suggestion
                  .mapM {
```